### PR TITLE
@gregori0o/algorithm kmeans

### PR DESCRIPTION
**Pull request by gregori0o**
Implementation of k-means algorithm and steps visualization.
- This is basic version of this algorithm. For numeric column distance is calculating by absolute value of difference. For other columns I use discrete metrics. ToDo: 1) add better method to find initial centroids than random; 2) improve stop criterion.
- Visualization works on numeric data. I am open to suggestion about appearance of visualization. ToDo: 1) add visualization of other types of data.
- Algorithm Engine and Algorithm Setup Widget is very simple. ToDo: 1) add some feature toggle of new algorithm to widget; 2) prepare engine and widget to easy adding new algorithm.
- ToDo: 1) Add QMessage Box and Loading window. 2) Add legend in clustering visualization 3) Move visualization setup to algorithm setup.

I'm going to realize ToDo.

### CODE REVIEW

**Review by t0maboro — APPROVED** [review]
In my opinion 🔥🔥

File: src/visualization/clustering/k_means_vis.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        except ValueError:
+            return False
+
+    def get_samples(self) -> List[int]:
+        array = np.arange(self.data.shape[0])
+        np.random.shuffle(array)
+        return list(array[:self.num_samples])
+
+    def click_listener(self, button_type: str):
+        if button_type == 'new_samples':
```
Comment (mhawryluk) [inline src/visualization/clustering/k_means_vis.py]: _match button_type:_

File: src/widgets/algorithm_setup_widget.py [hunk 2]
```diff
@@ -14,42 +17,94 @@ def __init__(self, parent, engine):
...
-        self.num_clusters_label_2 = QLabel(self.options_group)
-        self.num_clusters_label_2.setGeometry(QRect(20, 160, 131, 51))
-        self.num_clusters_label_2.setAlignment(Qt.AlignCenter)
-        self.num_clusters_label_2.setText("initialization method")
+        # dialog error
+        self.dialog = QDialog(self)
+        self.dialog.setFixedHeight(50)
+        self.dialog.setFixedWidth(400)
+        self.dialog.setWindowTitle("Error message")
+        QLabel("You didn't set data.\nPlease return to 'IMPORT DATA' window.", self.dialog)
```
Comment (mhawryluk) [inline src/widgets/algorithm_setup_widget.py:99]:
```suggestion
        QLabel("Data not set.\nPlease return to the 'IMPORT DATA' window.", self.dialog)
```
\+ turn that into QMessageBox to match what Tomek did in PreprocessingWidget::get_data

File: src/widgets/algorithm_run_widget.py [hunk 3]
```diff
@@ -14,25 +17,85 @@ def __init__(self, parent, engine):
...
+        # step by step
+        self.steps_group = QGroupBox()
+        self.steps_group.setTitle("Step by step")
+        self.steps_group.setMinimumWidth(200)
+        self.steps_group_layout = QFormLayout(self.steps_group)
+
+        # "run" button
+        self.run_steps_button = QPushButton("RUN")
+        self.run_steps_button.clicked.connect(partial(self.click_listener, 'run_steps'))
+        self.steps_group_layout.addRow(self.run_steps_button)
```
Comment (mhawryluk) [inline src/widgets/algorithm_run_widget.py:46]: I'm not a fan of having to click _run_ in the AlgorithmRunWidget and then having the actual visualization panel appear. I would move those initial animation/step_by_step options into AlgorithmSetupWidget. Also clicking the _submit and run_ button could automatically unfold the next widget. I can implement that as part of my layout revamp though.

File: src/visualization/clustering/k_means_vis.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+        self.oy_box = QComboBox()
+        self.oy_box.addItems(columns)
+        self.settings_box_layout.addRow(QLabel("OX:"), self.ox_box)
+        self.settings_box_layout.addRow(QLabel("OY:"), self.oy_box)
+
+        self.axis_button = QPushButton("Submit axis")
+        self.axis_button.clicked.connect(partial(self.click_listener, 'set_axis'))
+        self.settings_box_layout.addRow(self.axis_button)
+
+        # visualization layout
```
Comment (mhawryluk) [inline src/visualization/clustering/k_means_vis.py:121]: To do in the future: add legend widget to explain visualization symbols, for example what a square and a circle represent.

File: src/visualization/clustering/k_means_vis.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        self.axes.set_ylabel(name_y)
+        self.axes.set_xlim(min_x, max_x)
+        self.axes.set_ylim(min_y, max_y)
+        self.draw()
+
+    def new_centroids_plot(self, old_vector_x_centroids, old_vector_y_centroids, vector_x_centroids, vector_y_centroids,
+                           name_x, name_y, min_x, max_x, min_y, max_y):
+        self.axes.cla()
+        max_label = len(vector_x_centroids)
+        if not old_vector_x_centroids is None:
```
Comment (mhawryluk) [inline src/visualization/clustering/k_means_vis.py]:
```suggestion
        if old_vector_x_centroids is not None:
```

File: src/widgets/algorithm_run_widget.py [hunk 6]
```diff
@@ -14,25 +17,85 @@ def __init__(self, parent, engine):
...
+        horizontal_layout.addWidget(self.animation_group)
+        horizontal_layout.addStretch(1)
+        horizontal_layout.addWidget(self.steps_group)
+        horizontal_layout.addStretch(3)
+        self.layout.addLayout(horizontal_layout)
+        self.layout.addStretch(3)
+        self.layout.update()
 
+    def click_listener(self, button_type: str):
+        if button_type == 'run_steps':
```
Comment (mhawryluk) [inline src/widgets/algorithm_run_widget.py:88]: _match_ could be used, but it does add one more indentation level, so we can omit that

**Review by mhawryluk — APPROVED** [review]
Good job :) 
I did leave a few comments, but mostly suggestions for the future.

**Pull request merged**